### PR TITLE
#1576: Add target collateral ratio to orderbook calculations

### PR DESCRIPTION
**Pull request by svk31**
@abitmore I'd appreciate if you could look this over since I'm not sure it's correct. As mentioned before this stuff makes my brain hurt lol.

### CODE REVIEW

File: app/lib/common/MarketClasses.js [hunk 1]
```diff
@@ -721,12 +733,92 @@ class CallOrder {
...
+        /*
+        BSIP38:
+        * max_amount_to_sell = (debt * target_CR - collateral * feed_price)
+        * / (target_CR * match_price - feed_price)
+        */
+        if (this.target_collateral_ratio) {
+            let feed_price = this._getFeedPrice();
+            let match_price = this._getMatchPrice();
+            let nominator =
+                this.debt * this.target_collateral_ratio -
```
Comment (abitmore) [inline app/lib/common/MarketClasses.js]: Make sure `target_collateral_ratio` here (and same below) has been divided by 1000, for example, when the value is `1750`, need to use `1.75`.

File: app/lib/common/MarketClasses.js [hunk 2]
```diff
@@ -721,12 +733,92 @@ class CallOrder {
...
+                this.collateral * feed_price;
+            let denominator =
+                this.target_collateral_ratio * match_price - feed_price;
+            return nominator / denominator;
```
Comment (abitmore) [inline app/lib/common/MarketClasses.js]: I don't know if you've handled precision difference. I guess the value returned in `else` below is in Satoshi's, and what unit is returned here?

File: app/lib/common/MarketClasses.js [hunk 3]
```diff
@@ -721,12 +733,92 @@ class CallOrder {
...
+        BSIP38:
+        * max_amount_to_sell = (debt * target_CR - collateral * feed_price)
+        * / (target_CR * match_price - feed_price)
+        */
+        if (this.target_collateral_ratio) {
+            let feed_price = this._getFeedPrice();
+            let match_price = this._getMatchPrice();
+            let nominator =
+                this.debt * this.target_collateral_ratio -
+                this.collateral * feed_price;
```
Comment (abitmore) [inline app/lib/common/MarketClasses.js]: And Make sure `this.collateral * feed_price` has same unit as `this.debt` above.

File: app/lib/common/MarketClasses.js [hunk 4]
```diff
@@ -762,8 +859,9 @@ class CallOrder {
         if (newOrder.borrowers.indexOf(order.borrower) === -1) {
             newOrder.borrowers.push(order.borrower);
         }
-        newOrder.to_receive += order.to_receive;
-        newOrder.for_sale += order.for_sale;
+        newOrder.debt += order.debt;
+
+        newOrder.collateral += order.collateral;
```
Comment (abitmore) [inline app/lib/common/MarketClasses.js]: I don't know whether need to use 'max_collateral_for_sale' here.

File: app/lib/common/MarketClasses.js (hunk 1 above)
Comment (svk31) [reply app/lib/common/MarketClasses.js]: Right I remember wondering if I had to do that. Right now it's being used as the full int value, 1750, and all amounts are in satoshis. The prices are also adjusted to be useable with satoshis in _getFeedPrice and _getMatchPrice.

So even when using satoshi amounts the ratio should be divided by 1000?

File: app/lib/common/MarketClasses.js (hunk 2 above)
Comment (svk31) [reply app/lib/common/MarketClasses.js]: Should be satoshis also

File: app/lib/common/MarketClasses.js (hunk 3 above)
Comment (svk31) [reply app/lib/common/MarketClasses.js]: Yea that's why I'm inverting the feed and match prices if necessary.

File: app/lib/common/MarketClasses.js (hunk 4 above)
Comment (svk31) [reply app/lib/common/MarketClasses.js]: It gets pretty messy when summing orders now, I'll have to think about that one a bit (no pun intended).

File: app/lib/common/MarketClasses.js (hunk 1 above)
Comment (abitmore) [reply app/lib/common/MarketClasses.js]: Yes. Or multiply the other number by 1000, but also need to do multiplication for the denominator.

File: app/lib/common/MarketClasses.js (hunk 3 above)
Comment (abitmore) [reply app/lib/common/MarketClasses.js]: Has the same `/1000` issue as well.

File: app/lib/common/MarketClasses.js (hunk 1 above)
Comment (svk31) [reply app/lib/common/MarketClasses.js]: Is the number supposed to be negative?

**Comment by svk31** [comment]
Ok I **think** I have this right now!

**Pull request merged**